lib/puppet/provider/aix_object.rb

```diff
@@ -0,0 +1,390 @@
+    end
+
+    # Add a mapping from a Puppet property to an AIX attribute. The mapping_info must include:
+    #
+    #   * :puppet_property       -- The puppet property corresponding to this attribute
+    #   * :aix_attribute         -- The AIX attribute corresponding to this attribute. Defaults
+    #                            to puppet_property if this is not provided.
+    #   * :property_to_attribute -- A lambda that converts a Puppet Property to an AIX attribute
+    #                            value. Defaults to the identity function if not provided.
+    #   * :attribute_to_property -- A lambda that converts an AIX attribute to a Puppet property.
+    #                            Defaults to the identity function if not provided.
+    def mapping(mapping_info = {})
+      identity_fn = lambda { |x| x }
+      mapping_info[:aix_attribute] ||= mapping_info[:puppet_property]
+      mapping_info[:property_to_attribute] ||= identity_fn
+      mapping_info[:attribute_to_property] ||= identity_fn
+
+      mappings[:aix_attribute][mapping_info[:puppet_property]] = MappedObject.new(
+        mapping_info[:aix_attribute],
+        :convert_property_value,
+        mapping_info[:property_to_attribute]
+      )
+      mappings[:puppet_property][mapping_info[:aix_attribute]] = MappedObject.new(
+        mapping_info[:puppet_property],
+        :convert_attribute_value,
+        mapping_info[:attribute_to_property]
+      )
+    end
+
+    # Creates a mapping from a purely numeric Puppet property to
+    # an attribute
+    def numeric_mapping(mapping_info = {})
+      property = mapping_info[:puppet_property]
+
+      # We have this validation here b/c not all numeric properties
+      # handle this at the property level (e.g. like the UID). Given
+      # that, we might as well go ahead and do this validation for all
+      # of our numeric properties. Doesn't hurt.
+      mapping_info[:property_to_attribute] = lambda do |value|
+        unless value.is_a?(Integer)
+          raise ArgumentError, _("Invalid value #{value}: #{property} must be an Integer!")
+        end
+
+        value.to_s
+      end
+
+      # AIX will do the right validation to ensure numeric attributes
+      # can't be set to non-numeric values, so no need for the extra clutter.
+      mapping_info[:attribute_to_property] = lambda { |x| x.to_i }
+
+      mapping(mapping_info)
+    end
+
+    #-------------
+    # Useful Class Methods
+    # ------------
+
+    # Defines the getter and setter methods for each Puppet property that's mapped
+    # to an AIX attribute. We define only a getter for the :attributes property.
+    #
+    # Provider subclasses should call this method after they've defined all of
+    # their <puppet_property> => <aix_attribute> mappings.
+    def mk_resource_methods
+      # Define the Getter methods for each of our properties + the attributes
+      # property
+      properties = [:attributes]
+      properties += mappings[:aix_attribute].keys
```

Comment:
Seems like code smell to me to have all this generic code (dynamically created setters/getters, for example) mixed into the same file as AIX-specific code. Is this code at all useful at a more generic level, for other platforms?

Reply:
It's a common idiom in Puppet types/providers to do this. See for example:

https://github.com/puppetlabs/puppet/blob/5.5.x/lib/puppet/provider/user/directoryservice.rb#L25
https://github.com/puppetlabs/puppet/blob/5.5.x/lib/puppet/provider/parsedfile.rb#L162
https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider.rb#L414

The reason is b/c a property's `retrieve` and assignment methods delegate to these getters and setters first before anything else -- see https://github.com/puppetlabs/puppet/blob/master/lib/puppet/property.rb#L470 and https://github.com/puppetlabs/puppet/blob/master/lib/puppet/property.rb#L506 and https://github.com/puppetlabs/puppet/blob/master/lib/puppet/property.rb#L188

Personally, I like this kind of meta-programming. It's a nice, clean way of capturing the high level behavior that's going on, especially in this class where we see that all properties mapping to AIX attributes will be retrieved and set the same way. For this case, I do not think this is a code-smell.

Reply:
How much of the code in this file is only useful for AIX? I'd like this file to be as small as possible, so we can see what is truly different from platform to platform in files like these. (Essentially, make files matching "*aix*" as straightforward as possible, and move any meta-programming code into a different file that other platforms can use.)

Reply:
In other words, could any of this code move up into `Puppet::Provider`?

Reply:
Not in an obvious way, no. Puppet::Provider already has a default `mk_resource_methods` implemention here https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider.rb#L414 which is different from how we're using it.